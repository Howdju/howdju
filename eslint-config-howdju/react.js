module.exports = {
  parserOptions: {
    "ecmaFeatures": {
      "jsx": true
    },
  },
  env: {
    "browser": true,
  },
  extends: [
    "plugin:react/recommended"
  ],
  plugins: [
    "react",
  ],
  rules: {
    indent: [
      "warn",
      2,
      {
        "SwitchCase": 1,
        "ignoredNodes": [
          // We indent to the indent of the first prop, which this rule doesn't support
          "JSXAttribute",
          "JSXSpreadAttribute",
        ],
      }
    ],
    "react/boolean-prop-naming": "error",
    "react/no-children-prop": [
      // react-md requires us to use a children prop
      "off",
    ],
    "react/prop-types": [
      // This rule can't pick up props added by redux's mapStateToProps/mapDispatchToProps
      "off",
    ],
    // This doesn't seem to help; need to disable indent on JSXAttribute
    // "react/jsx-indent-props": [
    //   // We indent to the indent of the first prop, which this rule doesn't support
    //   "off"
    // ],
    "require-yield": [
      // Some of our sagas are side-effect only, requiring no yields.  And I don't think there's any harm in this.
      "off"
    ],
  },
  settings: {
    react: {
      version: "detect",
    },
  },
};
