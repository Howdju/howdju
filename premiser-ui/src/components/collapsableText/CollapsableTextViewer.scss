.collapsable-text-ellipsis {
  font-weight: bold;
}
