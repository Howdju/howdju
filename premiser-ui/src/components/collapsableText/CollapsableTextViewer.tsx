import React, { useState, HTMLAttributes } from "react";
import { Button } from "react-md";

import { truncateText } from "@/viewModels";

import "./CollapsableTextViewer.scss";

// two or more line breaks will indicate a paragraph
const paragraphBreak = /(\r\n|\n|\r){2,}/g;
const onlyWhitespace = /^\s*$/;

interface Props extends HTMLAttributes<HTMLDivElement> {
  text: string | undefined;
  /** If text is longer than this it will be truncated when collapsed. */
  truncateLength?: number;
}

export default function CollapsableTextViewer({
  text,
  truncateLength = 256,
  ...rest
}: Props) {
  const [isExpanded, setIsExpanded] = useState(false);

  const isTextLong = (text?.length ?? 0) > truncateLength;
  const displayText =
    !isTextLong || isExpanded
      ? text
      : truncateText(text, { omission: "", length: truncateLength });
  const paragraphs = makeParagraphs(displayText);

  const expandEllipsis = (
    <span
      className="collapsible-text-ellipsis clickable"
      onClick={() => setIsExpanded(true)}
    >
      â€¦
    </span>
  );

  return (
    <div {...rest}>
      <div>
        {paragraphs?.map((para, i) => {
          const key = `quote-text-para-${i}`;
          if (isTextLong && !isExpanded && i === paragraphs.length - 1) {
            return (
              <p key={key}>
                {para} {expandEllipsis}
              </p>
            );
          }
          return <p key={key}>{para}</p>;
        })}
      </div>
      {isTextLong && !isExpanded && (
        <Button
          flat
          children="More"
          className="text-expand-toggle"
          onClick={() => setIsExpanded(true)}
        />
      )}
      {isTextLong && isExpanded && (
        <Button
          flat
          children="Less"
          className="text-expand-toggle"
          onClick={() => setIsExpanded(false)}
        />
      )}
    </div>
  );
}

function makeParagraphs(quote: string | undefined) {
  return quote
    ?.split(paragraphBreak)
    .filter((para) => !onlyWhitespace.test(para));
}
