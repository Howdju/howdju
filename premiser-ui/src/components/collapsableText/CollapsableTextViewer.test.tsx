import React from "react";
import { render } from "@testing-library/react";
import dedent from "dedent";

import CollapsableTextViewer from "./CollapsableTextViewer";

describe("CollapsableTextViewer", () => {
  test("renders text containing markdown", () => {
    const text = dedent`
      paragraph 1

      paragraph 2

      | table  | header |
      | ------ | ------ |
      | cell1  | cell2  |
      | cell3  | cell4  |
      `;
    const { container } = render(<CollapsableTextViewer text={text} />);
    expect(container).toMatchSnapshot();
  });
});
