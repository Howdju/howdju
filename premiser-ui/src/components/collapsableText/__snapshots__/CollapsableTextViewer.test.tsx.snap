// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsableTextViewer renders text containing markdown 1`] = `
<div>
  <div
    class="collapsible-text-viewer"
  >
    <p>
      paragraph 1
    </p>
    

    <p>
      paragraph 2
    </p>
    

    <table>
      <thead>
        <tr>
          <th>
            table
          </th>
          <th>
            header
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            cell1
          </td>
          <td>
            cell2
          </td>
        </tr>
        <tr>
          <td>
            cell3
          </td>
          <td>
            cell4
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
